### DISCUSSION TITLE
Let's use `mocha --dry-run` to discover tests and remove esbuild

### DISCUSSION THREAD

**Discussion by muhammadyusuf-kurbonov** [CoderLine/mocha-vscode#172] [Ideas]
Let's use `mocha --dry-run` to discover tests and remove esbuild

Reasons:
1. Single responsibility principle: extension is needed only for running mocha
2. Esbuild is making some bugs (esbuild installation had issue, source-map is sometimes wrong)
3. Easy to discover avoiding custom logic (no need to parse files and source code)

**Comment by Danielku15** [comment]
I rather see negative points in changing (fully) to such a discovery mode. 

> mocha --dry-run

I remember problems with using `--dry-run` that various calls were not made to the reporter or information not being provided which is required to really build this extension. (more information below)

> Single responsibility principle: extension is needed only for running mocha

I don't agree with this simplification of what this extension does. This extension aims to be a full **integration** of Mocha into VSCode. This goes beyond just running mocha. 

Mocha is lacking of good (and lightweight) interfaces to load tests with the required metadata. The problem already starts with mapping the exact locations of tests in the source code. There is no information about the lines in which a test is placed breaking all functionality to jump to the tests from the test explorer and results. 

> Esbuild is making some bugs (esbuild installation had issue, source-map is sometimes wrong)

Assuming we discover tests through other means, we still need some level of source mapping. Mocha does not provide a semantic result on where an error happened. Currently the `Error` and its stack have to be inspected for this. 

Bugs with integrations can always happen and this extension is young, but as tooling evolves there will always be challenges unless standards are defined. We have tsx, ts-node and also Node added basic type annotation support lately. 

> Easy to discover avoiding custom logic (no need to parse files and source code)

As mentioned above, some complexities remain or are only shifted. We still need certain integration points with mocha that go further. 

-----

I am not saying `mocha --dry-run` might not work at all but my experience tells me that we only have some benefits in costs of others. Luckily we are not force to one solution but can offer some flexibility. This extension already supports 3 modes of discovering tests, all with their pros and cons. 

If you are eager to give it a try, you can horizontally extend the discoverer options and add a `mocha-dry-run` discoverer. I'm definitely not against adding one and if it proofs to be a great alternative it could become the default in future. 

https://github.com/CoderLine/mocha-vscode/blob/6bf76562638a8365eb952d23fae1bd5dc3451ead/src/discoverer/settings.ts#L35-L48

**Reply by muhammadyusuf-kurbonov** [reply]
Thanks for your response. Yes, I have tried to implement it. No success (yet) 🥲🥲🥲. Mocha doesn't return the line of the test or the exact stack trace of the error. It was just a solution (I didn't mean to offend you). So, I will open a PR at Mocha asking to provide this data. Also, I started to review other IDE extensions (Jetbrans NodeJS, emacs, slimeText), and other VSCode extensions (of Jest, Mocha sidebar, etc) to find some inspiration and "steal" ideas. Thanks for your work and have a nice work week :)
